File: 0141.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    struct ListNode *fast = head;
    struct ListNode *slow = head;
    
    if (fast != NULL) {
        fast = head->next;
    }
    
    while (slow && fast) {
        if (slow == fast) {
            return true;
        }
        fast = fast->next;
        if (fast) {
            fast = fast->next;
        }
        slow = slow->next;
    }
    
    return false;
}

File: 0141.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        it = head
        s = set([])
        
        while it != None:
            if it in s:
                return True
            else:
                s.add(it)
            it = it.next
        return False
            